File: Stage06-FromFile/Lua/main.lua
--[[
Play() loop
1. exits = here.displayLocation() # gets exits list, so include exit name
2. exits is passed to takeAction()
3. takeAction() displays exits as part of the menu
4. menu returns "Go "* if an exit is chosen
5. direction currently gets exit from string splicing "Go direction" to "direction" (north, east, south, west)
6. if exit contained "Go east -> coridoor" improve string splicing to separate "east" and "coridoor"

location.displayLocation()
1. return exit name for each exit: "east -> coridoor","north -> magic portal"
]]

local kb = require"lib.kboard"
local Console = require "lib.console"
local Narrator = require "narrator"
local Player = require "player"
local Shared = require "shared"
local Game = require "game"
--[[
global variables needed are dictionaries of Items, Enemies and Locations.
All global variables are stored in 'Shared.lua' and prefixed with Shared.
e.g. Shared.dictItems = {} holds a dictionary of Item objects available to any file via 'require("Shared")
]]

string.Split = Shared.Split
string.Trim = Shared.Trim
table.IsEmpty = Shared.IsEmpty

local function CheckExit(here, direction)
	--[[ check if next location has an item_required to enter ]]
	local nextLocation
	if direction == "north" then
		nextLocation = Shared.Locations[here:GetToNorth()]
	elseif direction == "east" then
		nextLocation = Shared.Locations[here:GetToEast()]
	elseif direction == "south" then
		nextLocation = Shared.Locations[here:GetToSouth()]
	elseif direction == "west" then
		nextLocation = Shared.Locations[here:GetToWest()]
	end	
	
	local itemRequired = nextLocation:GetItemRequired()	
	if itemRequired ~= "" then					-- key or other item needed
		if Shared.InList(Player.Inventory, itemRequired) then -- is it in player inventory?
			print("You use the "..itemRequired.." in your inventory to enter...")
		else
			print("You need "..itemRequired.." in your inventory to go that way")
			nextLocation = here
		end
		Shared.Sleep(2)
	end
	
	return nextLocation:GetName()	
end

local function Fight(here)
	--[[ pick a fight with an enemy ]]
	local enemy = Shared.Enemies[here:GetEnemy()]	-- use variable for convenience
	print(Player.Attack(here))						-- player.attack() returns a string message
	local message = ""
	if enemy:GetHealth() == 0 then
		message = "The "..enemy:GetName().." is dead"
		if enemy:GetDropItem() ~= "" then
			here:AddItem(enemy:GetDropItem())
			message = message .." and dropped a ".. enemy:GetDropItem()
		end
		here:SetEnemy("")	
	else
		message = enemy:Attack(Player)
	end
	print(message)									-- print attack details
	
	if Player.Health == 0 then
		print("You died at the hands of the "..enemy:GetName())
		Shared.Sleep(2)
		print("The game is over!")
		Shared.Sleep(2)
		Shared.Gamestate = Shared.Gamestates['dead']-- this breaks the gameloop in main()
	end
end	

local function TakeAction(here, exits, row)
	--[[ choose player action ]]
	local options = {}
	-- check for enemies
	local enemy = here:GetEnemy()
	if enemy ~= "" then
		table.insert(options, "Attack "..enemy.. "!")	-- add "Attack..." to options
	end
	-- examine / take any items
	local items = here:GetItems()
	if #items > 0 then
		for _, item in ipairs(items) do
			table.insert(options, "Examine "..item)
			table.insert(options, "Take "..item)
		end
	end
	-- open inventory
	if #Player.Inventory > 0 then
		table.insert(options, "Open your inventory")
	end
	-- take an exit
	if #exits > 0 then
		for _, exit in pairs(exits) do
			table.insert(options, "Go "..exit)
		end
	end
	
	-- quit the game	
	table.insert(options, "Quit")
	local choice = options[kb.menu("What next?", options, row)]
	-- choice examples: "Attack...", Examine...", "Go..." etc
	if choice:find("Attack") ~= nil then
		return "fight",""	
	elseif choice:find("Quit") ~= nil then
		return "quit", ""
	elseif choice:find("Go ") ~= nil then
		-- exit format "Go east -> coridoor" needs to be parsed differently
		local data = choice:Split("->")			-- "Go east -> coridoor" -> "Go east ", " corridoor"
		local direction = data[1]:sub(4):Trim() -- "Go east " -> "east"
		return "go", direction
	elseif choice:find("Examine") ~= nil then
		return "examine", choice:sub(9)
	elseif choice:find("Take") ~= nil then
		return "take", choice:sub(6)
	elseif choice:find("inventory") ~= nil then
		-- open inventory
		return "inventory", ""
	end
end

local function UseInventory()
	local choice = ""
	while choice ~= "Exit menu" do
		local row = Console.Clear()
		local title = "Choose your option:"
		local options = {}
		for _, item in pairs(Player.Inventory) do
			table.insert(options, "Examine "..item)
			table.insert(options, "Drop "..item)
		end
		table.insert(options, "Exit menu")
		choice = options[kb.menu(title, options, row)]
		if choice:find("Examine") ~= nil then
			local item = choice:sub(9)
			row = Console.Clear()
			print("You examine the ".. item)
			print("It is "..Shared.Items[item]:GetDescription())
			if Shared.Items[item].type == "Weapon" then
				print("It is a Weapon with "..Shared.Items[item]:GetDamage().." hit points!")
			end
			Shared.Sleep(3)
		elseif choice:find("Drop") ~= nil then
			local item = choice:sub(6)
			row = Console.Clear()
			print("You drop the ".. item)
			Shared.Locations[Shared.CurrentLocation]:AddItem(item)
			Player.RemoveFromInventory(item)
			Shared.Sleep(3)
		end
		if #Player.Inventory == 0 then
			choice = "Exit menu"
		end
	end
end

local function Play()
	--[[ make choices about items in locations, inventory, move around etc ]]
	local action, param
	local row = Console.Clear()
	local here = Shared.Locations[Shared.CurrentLocation]	-- Location object
	local exits = {}
	exits, row = here:DisplayLocation() 					-- returns available exits and current console row
	if not Console.IsConsole then row = -1 end				-- prevent /033[7;0H in ZeroBrane
	action, param = TakeAction(here, exits, row)
	row = Console.Clear()
	if action == "go" then
		-- check if next location has item_required
		Shared.CurrentLocation = CheckExit(here, param)		
	elseif action == "examine" then
		print("You examine the "..param)
		print(Shared.Items[param]:GetDescription())
		if Shared.Items[param].type == "Weapon" then
			print("It is a Weapon with "..Shared.Items[param]:GetDamage().." hit points!")
		end
		Shared.Sleep(2)
	elseif action == "take" then
		print("You take the "..param.." and put it in your backpack")
		Player.AddToInventory(param)
		here:RemoveItem(param)
		Shared.Sleep(2)
	elseif action == "inventory" then
		-- this action will only appear if Player.Inventory is not empty
		UseInventory()
	elseif action == "fight" then
		Fight(here) -- gamestate set to 'dead' if failed to win
		Shared.Sleep(4)
	elseif action == "quit" then
		Shared.Gamestate = Shared.Gamestates['quit']
	end
end

local function main()
	--[[ Everything runs from here ]]
	Console.Resize(80, 25)
	local row = Console.Clear()
	if Shared.Debug then							-- allow disabling debug mode
		Shared.Debug = not kb.getBoolean("Debug mode is ON. Do you want to disable debug mode?", row)
		row = Console.Clear()
	end
	if Game.LoadGame() then						-- load from file or use hard coded game if none exist
		--[[ game loop ]]
		Shared.Gamestate = Shared.Gamestates['play']							-- start at 'play'
		while Shared.Gamestate < Shared.Gamestates['quit'] do					-- continue game while in 'menu', 'play' gamestate
			Play()																-- play the game. Gamestate can be changed in multiple places
		end
		Console.Clear()	
		if Shared.Gamestate == Shared.Gamestates['dead'] then
			for _, message in ipairs(Narrator.DeathMessage) do
				print(message)
			end
		end
		for _, message in ipairs(Narrator.EndMessage) do
			print(message)
		end
	else
		print("Please correct errors in game locations before trying again")
	end
	if Console.IsConsole then
		io.write("Enter to quit")
		io.read()
	end
end
-- Program runs from here:
main()

File: Stage06-FromFile/Python/main.py
import time
import lib.kboard as kb #note use of lib. to locte file inside a subdirectory called lib
import lib.console as console
import game, player, shared, location, weapon, narrator

'''
play() loop
1. exits:list = here.display_location() # gets exits list, so include exit name
2. exits is passed to take_action()
3. take_action() displays exits as part of the menu
4. menu returns "Go "* if an exit is chosen
5. direction currently gets exit from string splicing "Go direction" to "direction" (north, east, south, west)
6. if exit contained "Go east -> coridoor" improve string splicing to separate "east" and "coridoor"

location.display_location()
1. return exit name for each exit: "east -> coridoor","north -> magic portal"
'''	
def check_exit(here:location.Location, direction:str) -> str:
	''' check if next location has an item_required to enter '''
	item_required:str = ""
	next_location:location.Location = here
	
	if direction == "to_north":
		next_location = shared.locations[here.to_north]
	elif direction == "to_east":
		next_location = shared.locations[here.to_east]
	elif direction == "to_south":
		next_location = shared.locations[here.to_south]
	elif direction == "to_west":
		next_location = shared.locations[here.to_west]
		
	item_required = next_location.item_required	
	if item_required != "":						# key or other item needed
		console.clear()
		if item_required in player.inventory: 	# is it in player inventory?
			print(f"You use the {item_required} in your inventory to enter...")
		else:
			print(f"You need {item_required} in your inventory to go that way")
			next_location = here
		time.sleep(2)
	
	return next_location.name	

def fight(here:location.Location):
	''' pick a fight with an enemy '''
	enemy = shared.enemies[here.enemy]		# use variable for convenience
	print(player.attack(here))				# player.attack() returns a string message
	message = ""
	if enemy.health == 0:
		message = f"The {here.enemy} is dead"
		if enemy.drop_item != "":
			here.add_item(enemy.drop_item)
			message += f" and dropped a {enemy.drop_item}"
		here.enemy = ""	
	else:
		message = enemy.attack(player)
	print(message)							# print attack details
	
	if player.health == 0:
		shared.gamestate = shared.gamestates['dead']		# this breaks the gameloop in main()

def play() -> None:
	''' make choices about items in locations, inventory, move around etc '''
	action:str = ""
	console.clear()
	here:location.Location = shared.locations[shared.current_location]
	exits, row = here.display_location() # returns available exits and current console row
	action, param = take_action(here, exits, row)
	if action == "go":
		#check if next location has item_required
		shared.current_location = check_exit(here, param)		
	elif action == "examine":
		print(f"You examine the {param}:")
		print(f"{shared.items[param].get_description()}")
		if type(shared.items[param]) == weapon.Weapon:
			print(f"It is a Weapon with {shared.items[param].damage} hit points!")
		time.sleep(2)
	elif action == "take":
		print(f"You take the {param} and put it in your backpack")
		player.add_to_inventory(param)
		here.remove_item(param)
		time.sleep(2)
	elif action == "inventory":
		# this action will only appear if player.inventory is not empty
		use_inventory()
	elif action == "fight":
		fight(here) # gamestate set to 'dead' if failed to win
		time.sleep(4)
	elif action == "quit":
		shared.gamestate = shared.gamestates['quit']

def	take_action(here:location.Location, exits:list, row:int) -> str:
	''' choose player action '''
	options:list = []
	# check for enemies
	if here.enemy != "":
		options.append(f"Attack {here.enemy} !")	# add "Attack..." to options
	# examine / take any items
	if len(here.items)> 0:
		for item in here.items:
			options.append(f"Examine {item}")
			options.append(f"Take {item}")
	# open inventory
	if len(player.inventory) > 0:
		options.append("Open your inventory")		# add "Open your inventory" to options
	# take an exit
	if len(exits) > 0:
		for exit in exits:
			options.append(f"Go {exit}")

	# quit the game	
	options.append("Quit")
	choice:str = options[kb.menu("What next?", options, row)]
	# choice examples: "Attack...", Examine...", "Go..." etc
	if "Attack" in choice:
		return "fight",""	
	elif choice == "Quit":
		return "quit", ""
	elif "Go " in choice:
		# get direction 
		# direction = f"to_{choice[3:]}" # to_north, to_east, etc
		# exit format "Go east -> coridoor" needs to be parsed differently
		data = choice[3:].split("->")
		direction = f"to_{data[0].strip()}" # to_north, to_east, etc
		return "go", direction
	elif "Examine" in choice:
		# get item
		item = choice[8:]
		return "examine", item
	elif "Take" in choice:
		item = choice[5:]
		return "take", item
	elif "inventory" in choice:
		# open inventory
		return "inventory", ""
	
def use_inventory():
	choice = ""
	while choice != "Exit menu":
		row = console.clear()
		title = "Choose your option:"
		options = []
		for item in player.inventory:
			options.append(f"Examine {item}")
			options.append(f"Drop {item}")

		options.append("Exit menu")
		choice = options[kb.menu(title, options, row)]
		if "Examine" in choice:
			item = choice[8:]
			row = console.clear()
			print(f"You examine the {item}")
			print(f"It is {shared.Items[item].description}")
			if type(shared.items[item]) == weapon.Weapon:
				print(f"It is a Weapon with {shared.items[item].damage} hit points!")

			time.sleep(3)
		elif "Drop" in choice:
			item = choice[5:]
			row = console.clear()
			print(f"You drop the {item}")
			shared.locations[shared.current_location].add_item(item)
			player.remove_from_inventory(item)
			time.sleep(3)

		if len(player.inventory) == 0:
			choice = "Exit menu"


def main() -> None:
	''' Everything runs from here '''
	console.resize(80, 25, False)
	row = console.clear()	# If you do not use os.system('cls') before trying to set cursor position in kb methods it shows the escape codes!
	if shared.debug:														# allow disabling debug mode
		shared.debug = not kb.get_boolean("Debug mode is ON. Do you want to disable debug mode?", row)
		console.clear()
	
	if game.load_game():
		''' game loop '''
		shared.gamestate = shared.gamestates['play']							# start at 'play'
		while shared.gamestate == shared.gamestates['play']:					# continue game while in 'play' gamestate
			play()																# play the game. Gamestate can be changed in multiple places
		
		console.clear()	
		if shared.gamestate == shared.gamestates['dead']:
			for message in narrator.death_message:
				print(message)
			
		for message in narrator.end_message:
			print(message)
	else:
		print("Please correct errors in game locations before trying again")

	if console.is_console:
		input("Enter to quit")
	
main()